### DISCUSSION TITLE
External link for show_pdf_page

### DISCUSSION THREAD

**Discussion by Andreasox** [pymupdf/PyMuPDF#2991] [Q&A]
Is it possible to link to an external file using show_pdf_page?
The use case is to reference external graphics on an existing page without having to recreate the PDF

**Comment by JorjMcKie** [comment]
By design, `show_pdf_page` is **_incapable_** of duplicating any annotation. This cannot be changed, because the source page is converted to a so-called "Form XObject". These objects cannot have annotations.
Here, _**annotation**_ means annotation the sense of PDF, which hence includes links and form fields.

All you can do is recreating links yourself on the target page: every annotation must be a child of the page that shows it - the target page in our case.

You must write code that extracts the links of the source page `links = srcpage.get_links()`.
Then compute a matrix that converts the source (page) rectangle coordinates to those of the bbox on the target page.
Fortunately (😉), there is a convenience method `torect()` in the `Rect`/`IRect` classes.
So a snippet may look like this (please confirm details / typos ...):
```python
links = srcpage.get_links()
tarpage.show_pdf_page(bbox, srcdoc, srcpage.number)
mat = srcpage.rect.torect(bbox)  # compute transformation matrix
for link in links:  # recreate links on target page
    link["from"] *= mat  # convert to target coordinates
    tarpage.insert_link(link)
```
Please note:
1. obviously only works for external links
2. if this also involves rotations, the matrix `mat` must be modified accordingly

**Comment by JorjMcKie** [comment]
Ah, I just realized that I did not understand your request.
The answer is **_NO_**, cannot be done in the way you describe it.

**Comment by JorjMcKie** [comment]
You can only "import" source pages of a PDF - and then optionally only show sub-rectangles of it on the target.

**Comment by Andreasox** [comment]
Thanks for the explanation, so PDF is in essence static in this aspect

**Reply by JorjMcKie** [reply]
yes, correct

**Reply by JorjMcKie** [reply]
It would be easier to directly define a link in the target, that points to a certain page in some other file. I.e. use `LINK_GOTOR` kind.
